src/Avalonia.Styling/StyledElement.cs

```diff
@@ -773,34 +784,45 @@ private void DetachStyles()
                     EndBatchUpdate();
                 }
             }
-
-            _styled = false;
         }
 
         private void DetachStyles(IReadOnlyList<IStyle> styles)
         {
-            styles = styles ?? throw new ArgumentNullException(nameof(styles));
 
+            DetachStylesOverride(styles);
+        }
+
+        internal protected virtual void DetachStylesOverride(IReadOnlyList<IStyle> styles)
+        {
+            styles = styles ?? throw new ArgumentNullException(nameof(styles));
             if (_appliedStyles is null)
             {
                 return;
             }
-
-            var count = styles.Count;
-
-            for (var i = 0; i < count; ++i)
+            BeginBatchUpdate();
```

Comment:
BeginBatchUpdate and EndBatchUpdate in here looks like something not related to controls visibility.
Can it be one more reason of improved performance? It can be moved to separated PR, although I am not sure why previous code was missing it @grokys